chapter10/ajd/exercise4.pl

```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/perl
+
+use strict;
+use warnings;
+use Data::Dumper;
+use v5.14.2;
+
+my $path = $ARGV[0] || "/home/ajd";
+my $data = data_for_path( $path );
+#dump_data_for_path( $path, $data );
+better_dump_data_for_path( $path, $data );
+
+sub data_for_path {
+  my ( $path ) = @_;
+
+  if ( -f $path or -l $path ) {
+    return undef;
+  }
+  elsif ( -d $path ) {
+    my %directory;
+    opendir ( my $dir, $path) || die ( "Can't open $path: $!\n" );
+    my @names = grep !/^\./, readdir( $dir );
+    closedir ( $dir );
+
+    foreach my $name ( @names ) {
+      $directory{$name} = data_for_path( "$path/$name" );
+    }
+
+    return \%directory;
+  }
+  warn "$path is broken somehow\n";
+  return undef;
+}
+
+sub dump_data_for_path {
+  my ( $path, $data ) = @_;
+
+  if ( !defined $data ) { # plain file
+    print "$path\n";
+    return;
+  }
+
+  foreach ( sort keys %$data ) {
+    dump_data_for_path( "$path/$_", $data->{$_} );
+  }
+}
+
+sub better_dump_data_for_path {
+  my ( $path, $data ) = @_;
+
+  state $depth = 0;
```

Comment:
Now you can only use `better_dump_data_for_path` once, or it'll continue to deeply nest results in another call. Not that it matters much in a script that calls it once (barring recursive calls).